File: code/packages/java/image-point-ops/src/test/java/com/codingadventures/imagepointops/ImagePointOpsTest.java
package com.codingadventures.imagepointops;

import com.codingadventures.pixelcontainer.PixelContainer;
import com.codingadventures.pixelcontainer.PixelOps;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class ImagePointOpsTest {

    /** Helper: a single-pixel container for tight checks. */
    private static PixelContainer single(int r, int g, int b, int a) {
        PixelContainer c = PixelOps.create(1, 1);
        PixelOps.setPixel(c, 0, 0, r, g, b, a);
        return c;
    }

    private static int[] px(PixelContainer c, int x, int y) {
        return PixelOps.pixelAt(c, x, y);
    }

    /* ---------------- u8-domain ---------------- */

    @Test
    void invertBasic() {
        PixelContainer out = ImagePointOps.invert(single(10, 20, 30, 128));
        assertArrayEquals(new int[]{245, 235, 225, 128}, px(out, 0, 0));
    }

    @Test
    void invertExtremes() {
        PixelContainer out = ImagePointOps.invert(single(0, 255, 128, 255));
        assertArrayEquals(new int[]{255, 0, 127, 255}, px(out, 0, 0));
    }

    @Test
    void invertPreservesDimensions() {
        PixelContainer src = PixelOps.create(4, 3);
        PixelContainer out = ImagePointOps.invert(src);
        assertEquals(4, out.width);
        assertEquals(3, out.height);
    }

    @Test
    void thresholdBelow() {
        PixelContainer out = ImagePointOps.threshold(single(50, 50, 50, 255), 128);
        assertArrayEquals(new int[]{0, 0, 0, 255}, px(out, 0, 0));
    }

    @Test
    void thresholdAbove() {
        PixelContainer out = ImagePointOps.threshold(single(200, 200, 200, 255), 128);
        assertArrayEquals(new int[]{255, 255, 255, 255}, px(out, 0, 0));
    }

    @Test
    void thresholdLuminanceIsPerceptual() {
        // Pure green is brighter perceptually (Y=0.7152*255≈182) than mean-equivalent.
        PixelContainer out = ImagePointOps.thresholdLuminance(single(0, 255, 0, 255), 150);
        assertArrayEquals(new int[]{255, 255, 255, 255}, px(out, 0, 0));
    }

    @Test
    void thresholdLuminanceBlueFails() {
        // Pure blue has very low luminance (Y=0.0722*255≈18).
        PixelContainer out = ImagePointOps.thresholdLuminance(single(0, 0, 255, 255), 50);
        assertArrayEquals(new int[]{0, 0, 0, 255}, px(out, 0, 0));
    }

    @Test
    void posterizeTwoLevels() {
        PixelContainer out = ImagePointOps.posterize(single(100, 200, 10, 255), 2);
        // step = 255, so each channel rounds to 0 or 255
        int[] p = px(out, 0, 0);
        for (int i = 0; i < 3; i++) assertTrue(p[i] == 0 || p[i] == 255);
    }

    @Test
    void posterizeIdentityAt256Levels() {
        PixelContainer out = ImagePointOps.posterize(single(42, 142, 242, 255), 256);
        assertArrayEquals(new int[]{42, 142, 242, 255}, px(out, 0, 0));
    }

    @Test
    void posterizeMinimumLevels() {
        // levels < 2 should clamp to 2.
        PixelContainer out = ImagePointOps.posterize(single(100, 200, 10, 255), 1);
        int[] p = px(out, 0, 0);
        for (int i = 0; i < 3; i++) assertTrue(p[i] == 0 || p[i] == 255);
    }

    @Test
    void swapRgbBgr() {
        PixelContainer out = ImagePointOps.swapRgbBgr(single(1, 2, 3, 4));
        assertArrayEquals(new int[]{3, 2, 1, 4}, px(out, 0, 0));
    }

    @Test
    void extractChannelRed() {
        PixelContainer out = ImagePointOps.extractChannel(single(10, 20, 30, 40), 0);
        assertArrayEquals(new int[]{10, 10, 10, 255}, px(out, 0, 0));
    }

    @Test
    void extractChannelGreen() {
        PixelContainer out = ImagePointOps.extractChannel(single(10, 20, 30, 40), 1);
        assertArrayEquals(new int[]{20, 20, 20, 255}, px(out, 0, 0));
    }

    @Test
    void extractChannelBlue() {
        PixelContainer out = ImagePointOps.extractChannel(single(10, 20, 30, 40), 2);
        assertArrayEquals(new int[]{30, 30, 30, 255}, px(out, 0, 0));
    }

    @Test
    void extractChannelAlpha() {
        PixelContainer out = ImagePointOps.extractChannel(single(10, 20, 30, 40), 3);
        assertArrayEquals(new int[]{40, 40, 40, 255}, px(out, 0, 0));
    }

    @Test
    void extractChannelInvalidReturnsZero() {
        PixelContainer out = ImagePointOps.extractChannel(single(10, 20, 30, 40), 9);
        assertArrayEquals(new int[]{0, 0, 0, 255}, px(out, 0, 0));
    }

    @Test
    void brightnessPositiveClamps() {
        PixelContainer out = ImagePointOps.brightness(single(200, 100, 50, 255), 100);
        assertArrayEquals(new int[]{255, 200, 150, 255}, px(out, 0, 0));
    }

    @Test
    void brightnessNegativeClamps() {
        PixelContainer out = ImagePointOps.brightness(single(200, 100, 50, 255), -100);
        assertArrayEquals(new int[]{100, 0, 0, 255}, px(out, 0, 0));
    }

    /* ---------------- linear-light ---------------- */

    @Test
    void contrastIdentityAtZero() {
        PixelContainer out = ImagePointOps.contrast(single(100, 128, 200, 255), 0);
        int[] p = px(out, 0, 0);
        // f=1 so output equals input for each channel.
        assertEquals(100, p[0]);
        assertEquals(128, p[1]);
        assertEquals(200, p[2]);
    }

    @Test
    void contrastIncreaseStretchesAwayFrom128() {
        PixelContainer out = ImagePointOps.contrast(single(100, 128, 200, 255), 100);
        int[] p = px(out, 0, 0);
        assertTrue(p[0] < 100);   // darks get darker
        assertEquals(128, p[1]);  // midpoint unchanged
        assertTrue(p[2] > 200);   // brights get brighter
    }

    @Test
    void gammaIdentity() {
        PixelContainer out = ImagePointOps.gamma(single(100, 150, 200, 255), 1.0);
        int[] p = px(out, 0, 0);
        for (int i = 0; i < 3; i++) assertTrue(Math.abs(p[i] - new int[]{100,150,200,255}[i]) <= 1);
    }

    @Test
    void gammaDarkenMidtones() {
        PixelContainer out = ImagePointOps.gamma(single(128, 128, 128, 255), 2.0);
        // g>1 in linear space darkens midtones.
        assertTrue(px(out, 0, 0)[0] < 128);
    }

    @Test
    void exposureZeroIsIdentity() {
        PixelContainer out = ImagePointOps.exposure(single(100, 150, 200, 255), 0);
        int[] p = px(out, 0, 0);
        for (int i = 0; i < 3; i++) assertTrue(Math.abs(p[i] - new int[]{100,150,200,255}[i]) <= 1);
    }

    @Test
    void exposurePositiveBrightens() {
        PixelContainer out = ImagePointOps.exposure(single(50, 50, 50, 255), 1.0);
        assertTrue(px(out, 0, 0)[0] > 50);
    }

    @Test
    void exposurePositiveClamps() {
        PixelContainer out = ImagePointOps.exposure(single(200, 200, 200, 255), 5.0);
        assertArrayEquals(new int[]{255, 255, 255, 255}, px(out, 0, 0));
    }

    @Test
    void greyscaleRec709() {
        // Pure red: Y = 0.2126, encoded back ~ small value.
        PixelContainer out = ImagePointOps.greyscale(single(255, 0, 0, 255), GreyscaleMethod.REC709);
        int[] p = px(out, 0, 0);
        assertEquals(p[0], p[1]);
        assertEquals(p[1], p[2]);
        assertTrue(p[0] > 50 && p[0] < 170);
    }

    @Test
    void greyscaleBt601DifferentFromRec709() {
        PixelContainer r1 = ImagePointOps.greyscale(single(255, 0, 0, 255), GreyscaleMethod.REC709);
        PixelContainer r2 = ImagePointOps.greyscale(single(255, 0, 0, 255), GreyscaleMethod.BT601);
        assertNotEquals(px(r1, 0, 0)[0], px(r2, 0, 0)[0]);
    }

    @Test
    void greyscaleAverage() {
        PixelContainer out = ImagePointOps.greyscale(single(255, 255, 255, 255), GreyscaleMethod.AVERAGE);
        assertEquals(255, px(out, 0, 0)[0]);
    }

    @Test
    void sepiaWhiteBecomesWarm() {
        PixelContainer out = ImagePointOps.sepia(single(255, 255, 255, 255));
        int[] p = px(out, 0, 0);
        // sepia matrix rows sum: r=1.351, g=1.203, b=0.937 → R clamps, G clamps, B doesn't.
        assertEquals(255, p[0]);
        assertEquals(255, p[1]);
        assertTrue(p[2] < 255);
    }

    @Test
    void colourMatrixIdentity() {
        double[][] id = { {1,0,0}, {0,1,0}, {0,0,1} };
        PixelContainer out = ImagePointOps.colourMatrix(single(100, 150, 200, 255), id);
        int[] p = px(out, 0, 0);
        for (int i = 0; i < 3; i++) assertTrue(Math.abs(p[i] - new int[]{100,150,200,255}[i]) <= 1);
    }

    @Test
    void saturateZeroGivesGrey() {
        PixelContainer out = ImagePointOps.saturate(single(200, 50, 10, 255), 0.0);
        int[] p = px(out, 0, 0);
        assertEquals(p[0], p[1]);
        assertEquals(p[1], p[2]);
    }

    @Test
    void saturateIdentityAtOne() {
        PixelContainer out = ImagePointOps.saturate(single(100, 150, 200, 255), 1.0);
        int[] p = px(out, 0, 0);
        for (int i = 0; i < 3; i++) assertTrue(Math.abs(p[i] - new int[]{100,150,200,255}[i]) <= 1);
    }

    @Test
    void hueRotate360IsIdentity() {
        PixelContainer out = ImagePointOps.hueRotate(single(200, 50, 10, 255), 360);
        int[] p = px(out, 0, 0);
        for (int i = 0; i < 3; i++) assertTrue(Math.abs(p[i] - new int[]{200,50,10,255}[i]) <= 2);
    }

    @Test
    void hueRotate180InvertsHue() {
        // Red rotated 180 → cyan-ish.
        PixelContainer out = ImagePointOps.hueRotate(single(255, 0, 0, 255), 180);
        int[] p = px(out, 0, 0);
        assertTrue(p[0] < p[1] || p[0] < p[2]);
    }

    @Test
    void hueRotateNegative() {
        // -120 should wrap to 240 and produce same result as +240.
        PixelContainer a = ImagePointOps.hueRotate(single(255, 0, 0, 255), -120);
        PixelContainer b = ImagePointOps.hueRotate(single(255, 0, 0, 255), 240);
        int[] pa = px(a, 0, 0), pb = px(b, 0, 0);
        for (int i = 0; i < 3; i++) assertTrue(Math.abs(pa[i] - pb[i]) <= 2);
    }

    @Test
    void hueRotateGreyscaleUnchanged() {
        // Saturation 0 → hue rotation should not change the colour.
        PixelContainer out = ImagePointOps.hueRotate(single(128, 128, 128, 255), 90);
        int[] p = px(out, 0, 0);
        assertEquals(p[0], p[1]);
        assertEquals(p[1], p[2]);
    }

    /* ---------------- whole-image sRGB ↔ linear ---------------- */

    @Test
    void srgbToLinearRoundTripApproximate() {
        PixelContainer src = single(100, 150, 200, 255);
        PixelContainer lin = ImagePointOps.srgbToLinearImage(src);
        PixelContainer back = ImagePointOps.linearToSrgbImage(lin);
        int[] p = px(back, 0, 0);
        // Quantisation through 8-bit midtones loses a little precision.
        for (int i = 0; i < 3; i++) assertTrue(Math.abs(p[i] - new int[]{100,150,200,255}[i]) <= 3);
    }

    /* ---------------- LUTs ---------------- */

    @Test
    void applyLut1dU8Identity() {
        byte[] lut = new byte[256];
        for (int i = 0; i < 256; i++) lut[i] = (byte) i;
        PixelContainer out = ImagePointOps.applyLut1dU8(single(10, 20, 30, 40), lut, lut, lut);
        assertArrayEquals(new int[]{10, 20, 30, 40}, px(out, 0, 0));
    }

    @Test
    void applyLut1dU8Inverts() {
        byte[] lut = new byte[256];
        for (int i = 0; i < 256; i++) lut[i] = (byte) (255 - i);
        PixelContainer out = ImagePointOps.applyLut1dU8(single(10, 20, 30, 40), lut, lut, lut);
        assertArrayEquals(new int[]{245, 235, 225, 40}, px(out, 0, 0));
    }

    @Test
    void buildLut1dU8Monotone() {
        byte[] lut = ImagePointOps.buildLut1dU8(x -> x); // identity in linear
        // Identity through decode/encode should approximately round-trip.
        for (int i = 0; i < 256; i++) {
            int v = lut[i] & 0xFF;
            assertTrue(Math.abs(v - i) <= 1, "LUT[" + i + "] off: " + v);
        }
    }

    @Test
    void buildGammaLutIsDecreasing() {
        byte[] lut = ImagePointOps.buildGammaLut(2.0);
        // Gamma 2.0 in linear darkens, so LUT values should mostly be below input.
        int below = 0;
        for (int i = 1; i < 256; i++) if ((lut[i] & 0xFF) <= i) below++;
        assertTrue(below > 200);
    }

    /* ---------------- multi-pixel correctness ---------------- */

    @Test
    void appliesToEveryPixel() {
        PixelContainer src = PixelOps.create(3, 2);
        for (int y = 0; y < 2; y++)
            for (int x = 0; x < 3; x++)
                PixelOps.setPixel(src, x, y, 10, 20, 30, 255);
        PixelContainer out = ImagePointOps.invert(src);
        for (int y = 0; y < 2; y++)
            for (int x = 0; x < 3; x++)
                assertArrayEquals(new int[]{245, 235, 225, 255}, px(out, x, y));
    }

    @Test
    void doesNotMutateInput() {
        PixelContainer src = single(10, 20, 30, 40);
        ImagePointOps.invert(src);
        assertArrayEquals(new int[]{10, 20, 30, 40}, px(src, 0, 0));
    }
}

File: code/packages/kotlin/image-point-ops/src/test/kotlin/com/codingadventures/imagepointops/ImagePointOpsTest.kt
package com.codingadventures.imagepointops

import com.codingadventures.pixelcontainer.PixelContainer
import com.codingadventures.pixelcontainer.PixelOps
import kotlin.math.abs
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertFails
import kotlin.test.assertTrue
import kotlin.test.assertNotEquals

class ImagePointOpsTest {

    // --- small helpers ---------------------------------------------------

    private fun pix(r: Int, g: Int, b: Int, a: Int = 255): PixelContainer {
        val c = PixelContainer(1, 1)
        PixelOps.setPixel(c, 0, 0, r, g, b, a)
        return c
    }

    private fun get(c: PixelContainer) = PixelOps.pixelAt(c, 0, 0)

    private fun assertNear(a: Int, b: Int, tol: Int = 2) {
        assertTrue(abs(a - b) <= tol, "expected ~$b got $a (tol=$tol)")
    }

    // --- invert ----------------------------------------------------------

    @Test fun invertInvertsRGB() {
        val out = ImagePointOps.invert(pix(10, 20, 30, 128))
        assertContentEquals(intArrayOf(245, 235, 225, 128), get(out))
    }

    @Test fun invertTwiceIsIdentity() {
        val once = ImagePointOps.invert(pix(50, 100, 150, 255))
        val twice = ImagePointOps.invert(once)
        assertContentEquals(intArrayOf(50, 100, 150, 255), get(twice))
    }

    // --- threshold -------------------------------------------------------

    @Test fun thresholdBelow() {
        val out = ImagePointOps.threshold(pix(10, 10, 10), 100)
        assertContentEquals(intArrayOf(0, 0, 0, 255), get(out))
    }

    @Test fun thresholdAbove() {
        val out = ImagePointOps.threshold(pix(200, 200, 200), 100)
        assertContentEquals(intArrayOf(255, 255, 255, 255), get(out))
    }

    @Test fun thresholdLuminanceWeightsGreenMost() {
        // pure green above threshold — white; pure red below — black
        val g = ImagePointOps.thresholdLuminance(pix(0, 200, 0), 100)
        val r = ImagePointOps.thresholdLuminance(pix(200, 0, 0), 100)
        assertEquals(255, get(g)[0])
        assertEquals(0, get(r)[0])
    }

    // --- posterize -------------------------------------------------------

    @Test fun posterizeTwoLevelsGivesBlackOrWhite() {
        val dark = ImagePointOps.posterize(pix(10, 10, 10), 2)
        val light = ImagePointOps.posterize(pix(250, 250, 250), 2)
        assertEquals(0, get(dark)[0])
        assertEquals(255, get(light)[0])
    }

    @Test fun posterizeRejectsOneLevel() {
        assertFails { ImagePointOps.posterize(pix(0, 0, 0), 1) }
    }

    // --- channel ops -----------------------------------------------------

    @Test fun swapRgbBgr() {
        val out = ImagePointOps.swapRgbBgr(pix(10, 20, 30))
        assertContentEquals(intArrayOf(30, 20, 10, 255), get(out))
    }

    @Test fun extractChannelRed() {
        val out = ImagePointOps.extractChannel(pix(100, 200, 50, 200), 0)
        assertContentEquals(intArrayOf(100, 100, 100, 255), get(out))
    }

    @Test fun extractChannelAlpha() {
        val out = ImagePointOps.extractChannel(pix(100, 200, 50, 77), 3)
        assertContentEquals(intArrayOf(77, 77, 77, 255), get(out))
    }

    @Test fun extractChannelRejectsBadIndex() {
        assertFails { ImagePointOps.extractChannel(pix(0, 0, 0), 4) }
    }

    // --- brightness / contrast ------------------------------------------

    @Test fun brightnessPositiveSaturates() {
        val out = ImagePointOps.brightness(pix(250, 250, 250), 50)
        assertContentEquals(intArrayOf(255, 255, 255, 255), get(out))
    }

    @Test fun brightnessNegativeSaturates() {
        val out = ImagePointOps.brightness(pix(10, 10, 10), -50)
        assertContentEquals(intArrayOf(0, 0, 0, 255), get(out))
    }

    @Test fun contrastOneIsIdentity() {
        val out = ImagePointOps.contrast(pix(40, 80, 200), 1.0)
        assertContentEquals(intArrayOf(40, 80, 200, 255), get(out))
    }

    @Test fun contrastZeroCollapsesToMidGrey() {
        val out = ImagePointOps.contrast(pix(0, 128, 255), 0.0)
        assertContentEquals(intArrayOf(128, 128, 128, 255), get(out))
    }

    // --- gamma / exposure -----------------------------------------------

    @Test fun gammaOneIsNearIdentity() {
        val out = ImagePointOps.gamma(pix(50, 100, 150), 1.0)
        val p = get(out)
        assertNear(p[0], 50); assertNear(p[1], 100); assertNear(p[2], 150)
    }

    @Test fun gammaDarkensWithG2() {
        val p = get(ImagePointOps.gamma(pix(128, 128, 128), 2.0))
        assertTrue(p[0] < 128)
    }

    @Test fun exposureZeroIsNearIdentity() {
        val out = ImagePointOps.exposure(pix(50, 100, 150), 0.0)
        val p = get(out)
        assertNear(p[0], 50); assertNear(p[1], 100); assertNear(p[2], 150)
    }

    @Test fun exposurePositiveBrightens() {
        val p = get(ImagePointOps.exposure(pix(50, 50, 50), 1.0))
        assertTrue(p[0] > 50)
    }

    @Test fun exposureNegativeDarkens() {
        val p = get(ImagePointOps.exposure(pix(200, 200, 200), -1.0))
        assertTrue(p[0] < 200)
    }

    // --- greyscale -------------------------------------------------------

    @Test fun greyscaleRec709PureWhite() {
        val p = get(ImagePointOps.greyscale(pix(255, 255, 255), ImagePointOps.GreyscaleMethod.REC709))
        assertNear(p[0], 255, 1); assertEquals(p[0], p[1]); assertEquals(p[1], p[2])
    }

    @Test fun greyscaleBt601GivesEqualChannels() {
        val p = get(ImagePointOps.greyscale(pix(200, 100, 50), ImagePointOps.GreyscaleMethod.BT601))
        assertEquals(p[0], p[1]); assertEquals(p[1], p[2])
    }

    @Test fun greyscaleAverageGivesEqualChannels() {
        val p = get(ImagePointOps.greyscale(pix(10, 20, 30), ImagePointOps.GreyscaleMethod.AVERAGE))
        assertEquals(p[0], p[1]); assertEquals(p[1], p[2])
    }

    // --- matrix / sepia --------------------------------------------------

    @Test fun sepiaShiftsTowardsBrown() {
        val p = get(ImagePointOps.sepia(pix(255, 255, 255)))
        // R > G > B is the sepia signature
        assertTrue(p[0] >= p[1] && p[1] >= p[2])
    }

    @Test fun identityColourMatrixIsIdentity() {
        val id = arrayOf(
            doubleArrayOf(1.0, 0.0, 0.0),
            doubleArrayOf(0.0, 1.0, 0.0),
            doubleArrayOf(0.0, 0.0, 1.0)
        )
        val p = get(ImagePointOps.colourMatrix(pix(50, 100, 150), id))
        assertNear(p[0], 50); assertNear(p[1], 100); assertNear(p[2], 150)
    }

    @Test fun colourMatrixRejectsBadShape() {
        val bad = arrayOf(doubleArrayOf(1.0, 0.0))
        assertFails { ImagePointOps.colourMatrix(pix(0, 0, 0), bad) }
    }

    // --- saturate --------------------------------------------------------

    @Test fun saturateZeroGivesGrey() {
        val p = get(ImagePointOps.saturate(pix(200, 50, 50), 0.0))
        assertEquals(p[0], p[1]); assertEquals(p[1], p[2])
    }

    @Test fun saturateOneIsNearIdentity() {
        val p = get(ImagePointOps.saturate(pix(100, 50, 20), 1.0))
        assertNear(p[0], 100); assertNear(p[1], 50); assertNear(p[2], 20)
    }

    // --- hue rotate ------------------------------------------------------

    @Test fun hueRotate360IsNearIdentity() {
        val p = get(ImagePointOps.hueRotate(pix(200, 100, 50), 360.0))
        assertNear(p[0], 200); assertNear(p[1], 100); assertNear(p[2], 50)
    }

    @Test fun hueRotate120PermutesPrimaries() {
        // red at 0° rotates to green at 120°
        val p = get(ImagePointOps.hueRotate(pix(255, 0, 0), 120.0))
        assertTrue(p[1] > p[0] && p[1] > p[2])
    }

    // --- srgb<->linear image --------------------------------------------

    @Test fun srgbToLinearRoundTripApproximate() {
        val src = pix(128, 64, 200)
        val lin = ImagePointOps.srgbToLinearImage(src)
        val back = ImagePointOps.linearToSrgbImage(lin)
        val p = get(back)
        assertNear(p[0], 128, 3); assertNear(p[1], 64, 3); assertNear(p[2], 200, 3)
    }

    // --- LUTs ------------------------------------------------------------

    @Test fun buildIdentityLutThenApplyIsNoOp() {
        val id = ImagePointOps.buildLut1dU8 { it }
        val out = ImagePointOps.applyLut1dU8(pix(10, 20, 30), id, id, id)
        val p = get(out)
        assertNear(p[0], 10, 1); assertNear(p[1], 20, 1); assertNear(p[2], 30, 1)
    }

    @Test fun gammaLutProducesReasonableValues() {
        val lut = ImagePointOps.buildGammaLut(2.0)
        assertEquals(0, lut[0].toInt() and 0xFF)
        assertEquals(255, lut[255].toInt() and 0xFF)
        // gamma 2.0 darkens mids
        assertTrue((lut[128].toInt() and 0xFF) < 128)
    }

    @Test fun applyLutRejectsShortLut() {
        assertFails {
            ImagePointOps.applyLut1dU8(pix(0,0,0), ByteArray(10), ByteArray(256), ByteArray(256))
        }
    }

    // --- multi-pixel shape ----------------------------------------------

    @Test fun operationsProduceSameShape() {
        val src = PixelContainer(3, 2)
        PixelOps.fillPixels(src, 10, 20, 30, 40)
        val out = ImagePointOps.invert(src)
        assertEquals(src.width, out.width)
        assertEquals(src.height, out.height)
    }

    @Test fun operationsDoNotMutateSource() {
        val src = pix(10, 20, 30, 40)
        val before = src.data.copyOf()
        ImagePointOps.invert(src)
        ImagePointOps.gamma(src, 2.0)
        assertContentEquals(before, src.data)
    }

    @Test fun invertVsOriginalDiffers() {
        val src = pix(10, 20, 30)
        assertNotEquals(
            get(src).toList(),
            get(ImagePointOps.invert(src)).toList()
        )
    }
}
